wgpu/utils/shadertoy.py

```diff
@@ -0,0 +1,311 @@
+    var out: Varyings;
+    if (index == u32(0)) {
+        out.position = vec4<f32>(-1.0, -1.0, 0.0, 1.0);
+        out.uv = vec2<f32>(0.0, 1.0);
+    } else if (index == u32(1)) {
+        out.position = vec4<f32>(3.0, -1.0, 0.0, 1.0);
+        out.uv = vec2<f32>(2.0, 1.0);
+    } else {
+        out.position = vec4<f32>(-1.0, 3.0, 0.0, 1.0);
+        out.uv = vec2<f32>(0.0, -1.0);
+    }
+    return out;
+
+}
+"""
+
+builtin_variables = """
+
+var<private> i_time: f32;
+var<private> i_resolution: vec2<f32>;
+var<private> i_time_delta: f32;
+var<private> i_mouse: vec2<f32>;
+var<private> i_frame: u32;
+
+// TODO: more global variables
+// var<private> i_frag_coord: vec2<f32>;
+
+"""
+
+fragment_code = """
+
+struct ShadertoyInput {
+    resolution: vec2<f32>,
+    mouse: vec2<f32>,
+    time: f32,
+    time_delta: f32,
+    frame: u32,
+};
+
+@group(0) @binding(0)
+var<uniform> input: ShadertoyInput;
+
+
+@stage(fragment)
+fn fs_main(in: Varyings) -> @location(0) vec4<f32> {
+
+    i_time = input.time;
+    i_resolution = input.resolution;
+    i_time_delta = input.time_delta;
+    i_mouse = input.mouse;
+    i_mouse.y = i_resolution.y - i_mouse.y;
+    i_frame = input.frame;
+
+
+    let uv = vec2<f32>(in.uv.x, 1.0 - in.uv.y);
+    let frag_coord = uv * i_resolution;
+
+    var fragColor = shader_main(frag_coord);
+
+    return vec4<f32>(fragColor.rgb, 1.0);
+}
+
+ """
+
+binding_layout = [
+    {
+        "binding": 0,
+        "visibility": wgpu.ShaderStage.FRAGMENT,
+        "buffer": {"type": wgpu.BufferBindingType.uniform},
+    }
+]
+
+uniform_dtype = [
+    ("resolution", "float32", (2)),
+    ("mouse", "float32", (2)),
+    ("time", "float32"),
+    ("time_delta", "float32"),
+    ("frame", "uint32"),
+    ("__padding", "uint8", (4)),  # padding to 32 bytes
+]
+
+
+class Shadertoy:
+
+    """
+    Parameters:
+        shader_code (str): The shader code to use.
+        resolution (tuple): The resolution of the shadertoy.
+
+    The shader code must contain a entry point function:
+
+    ``fn shader_main(frag_coord: vec2<f32>) -> vec4<f32>{}``
+
+    It has a parameter ``frag_coord`` which is the current pixel coordinate (in range 0..resolution, origin is bottom-left),
+    and it must return a vec4<f32> color, which is the color of the pixel at that coordinate.
+
+    some built-in variables are available in the shader:
+
+    * ``i_time``: the global time in seconds
+    * ``i_time_delta``: the time since last frame in seconds
+    * ``i_frame``: the frame number
+    * ``i_resolution``: the resolution of the shadertoy
+    * ``i_mouse``: the mouse position in pixels
+
+    """
+
+    # todo: add more built-in variables
+    # todo: support input textures
+    # todo: support multiple render passes (`i_channel0`, `i_channel1`, etc.)
+
+    def __init__(self, shader_code, resolution=(800, 450)) -> None:
+
+        self._uniform_data = np.zeros((), dtype=uniform_dtype)
+
+        self.shader_code = shader_code
```

Comment:
Let's store as `self._shader_code` and add a public `@property` getter. That way it's also listed in the documentation.